### PR TITLE
Support Block-Outlining in the Roslyn feature API.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide @Pilchie

**Comment by DustinCampbell** [comment]
The conflation of the vertical block lines with outlining (i.e. code-folding) confuses me. How is this supposed to work? Why are they intertwined?

File: src/EditorFeatures/Core/Implementation/Outlining/OutliningSpan.cs [hunk 1]
```diff
@@ -31,13 +32,22 @@ internal class OutliningSpan
         /// </summary>
         public bool IsDefaultCollapsed { get; }
 
-        public OutliningSpan(TextSpan textSpan, TextSpan hintSpan, string bannerText, bool autoCollapse, bool isDefaultCollapsed = false)
+        public string Type { get; }
+
+        public OutliningSpan(TextSpan textSpan, TextSpan hintSpan, string bannerText, bool autoCollapse, bool isDefaultCollapsed)
```
Comment (paulvanbrenk) [inline src/EditorFeatures/Core/Implementation/Outlining/OutliningSpan.cs]: Removing the default is a breaking change for TypeScript

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Outlining/OutliningSpan.cs]: It's not.  That API is left around for *binary* compat.  The new constructor will then be called if you recompile

**Comment by paulvanbrenk** [comment]
Looks good to me.

**Comment by CyrusNajmabadi** [comment]
> The conflation of the vertical block lines with outlining (i.e. code-folding) confuses me. How is this supposed to work? Why are they intertwined?

They are actually intertwined at the editor level (and that's a good thing IMO).  All structural-blocks are outlining blocks (because they can be collapsed).  i.e. whne you have structural blocks, you also get the editor outlining treatment on the margin.  

From the Roslyn perspective, we could keep these separate, but they would overlap on basicallly everything (which is why i ended up implementing it this way).

**Comment by CyrusNajmabadi** [comment]
Basically,the only difference between a structural block and the outlining block is that:

1. If you give a StructuralBlock a 'Type' that is not "NonStructural" then you get the vertical line treatment.
2. StructuralBlocks have the concept of 'Parent' for nesting levels.

**Comment by DustinCampbell** [comment]
@CyrusNajmabadi: Does that mean that cannot not have vertical lines without having collapsing in the margin? I would think these should be separate.

**Comment by DustinCampbell** [comment]
>If you give a StructuralBlock a 'Type' that is not "NonStructural" then you get the vertical line treatment.

I see

**Comment by CyrusNajmabadi** [comment]
> Does that mean that cannot not have vertical lines without having collapsing in the margin? I would think these should be separate.

Sorry, i misspoke.  You can choose to also not give it the outlining treatment if you don't want.  That's the IsCollapsible property.

**Comment by DustinCampbell** [comment]
Got it. Are we making any design changes to the existing collapsible blocks in this change? (I haven't explored it deeply yet.)

**Comment by CyrusNajmabadi** [comment]
We could have them separate.  But we'll have to have a lot of shared logic.  For example, we'll need to compute the exact same values for namespace outlining as we do for namespace block tagging.  etc. etc.  I found it much more palatable to say that htey're the same concept just with a few knobs.  For example, comments would be NonStructural-collapsible items.  Code-blocks would be Structural-non-collapsible items.  Existing outlining items (Except comments) would be structural-collapsible items.

**Comment by DustinCampbell** [comment]
I assuming using directives and imports are also NonStructural-collapsible items.

**Comment by CyrusNajmabadi** [comment]
The implementation also shows precisely how overlapping these features are.  The only deviations in the implementation are that the block provider returns a tag that just has a few more properties set.  That's it.  They share every other bit of code up to that point.

In this way, the data model is fully descriptive, but the host can decide the best treatment for the UI.  In the case of VS it's "give hte outlining margin treatment if it's collapsible, and give the vertical line treatment if it's Structural".  Given that these are just two bits of data (which are not even exclusive), it did not feel nearly different enough to warrant a different API to me. :)

**Comment by CyrusNajmabadi** [comment]
Note: from an API perspective, i think i'd want to rename things be more like IBlockSpanService and BlockSpan (from IOutliningService and OutliningSpan).  'BlockSpans' would be the currency we would use, and would have enough information to drive the VS experience here.

**Comment by DustinCampbell** [comment]
>Note: from an API perspective, i think i'd want to rename things be more like IBlockSpanService and BlockSpan (from IOutliningService and OutliningSpan). 'BlockSpans' would be the currency we would use, and would have enough information to drive the VS experience here.

That make sense.

**Comment by CyrusNajmabadi** [comment]
> I assume using directives and imports are also NonStructural-collapsible items.

Yup!

**Comment by CyrusNajmabadi** [comment]
> Got it. Are we making any design changes to the existing collapsible blocks in this change? (I haven't explored it deeply yet.)

No.  Right now you'd get exactly the same behavior as before.  (I always pass NonStructural for all blocks unless you give a specific type).  This just changes our internal currency so we *can* make our blocks structural in a followup PR.

**Comment by DustinCampbell** [comment]
Cool. Just wanted to be sure we had maximum flexibility. :smile:

File: src/EditorFeatures/Core/Implementation/Outlining/OutliningSpan.cs (hunk 1 above)
Comment (paulvanbrenk) [inline src/EditorFeatures/Core/Implementation/Outlining/OutliningSpan.cs]: Oh yeah... that's how that works... and not the other way around.

**Comment by CyrusNajmabadi** [comment]
Yup yup!

**Comment by CyrusNajmabadi** [comment]
> IOutliningService 

Correct.  And the only difference now is that you can pass in a new "Type" string into the OutliningSpans you produce.  Look at PredefinedStructureTypes2 for the types that are available.  (Note: this set is in flux and is being discussed with the editor team).  But basically, by default we pass "NonStructural" (which means 'no vertical line'.  If you pass a good value, you'll get a vertical line, and the color will be picked based on the Type you specify.  (that way the colors are consistent across languages).  

Note that the following are all sets of reasonable values for your OutliningSpan:

1. IsCollapsible=true, Type=NonStructural.  e.g. things like comments or imports.  Things you want to be able to collapse but which would look terrible with vertical bars because they're not indented.
2. IsCollapsible=false, Type=anythingelse.  e.g. things like if/for blocks (if you don't want to support users collapsing these), but you do want vertical bars.
3. IsCollapsible=true, Type=anythingelse.  e.g. for things like classes/namespaces that you want to both have outlining spans *and* vertical bars.

**Comment by CyrusNajmabadi** [comment]
@paulvanbrenk Would you be opposed to me making the change to call this BlockSpan now?  If so, i can def wait on that.

**Comment by paulvanbrenk** [comment]
> Would you be opposed to me making the change to call this BlockSpan now? If so, i can def wait on that.


Not against that, I believe we still have some time for preview 5. And it would make my implementation for TypeScript trivial.

If you're doing that you can remove the old constructor, since you'll break us anyway. 🎱

**Pull request merged**